OpenRA.Mods.Common/Traits/Buildings/Building.cs

```diff
@@ -204,12 +203,13 @@ public virtual bool IsCloseEnoughToBase(World world, Player p, string buildingNa
 					{
 						var unitsAtPos = world.ActorMap.GetActorsAt(pos).Where(a => a.IsInWorld
 							&& (a.Owner == p || (allyBuildEnabled && a.Owner.Stances[p] == Stance.Ally))
-							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>());
+							&& a.Info.HasTraitInfo<GivesBuildableAreaInfo>()
+							&& a.Info.TraitInfoOrDefault<GivesBuildableAreaInfo>().ValidTypes.Contains(AdjacentType));
```

Comment:
You should change the `TraitInfoOrDefault` here and below to only `TraitInfo`, as the `HasTraitInfo` checks already guarantee that the trait is there.